crates/sparrow-backend/src/logical_to_physical.rs

```diff
@@ -132,27 +150,32 @@ impl LogicalToPhysical {
                         let mut exprs = ExprPattern::new_input()?;
 
                         let data_type = self.reference_type(&arg)?.clone();
-                        exprs.add_instruction(
-                            "fieldref",
-                            // Note: that `merge` should produce a record with a
-                            // field for each merged step, identified by the
-                            // step ID being merged. This may change depending
-                            // on (a) optimizations that change steps and
-                            // whether it is difficult to update these and (b)
-                            // how we choose to implement merge.
-                            //
-                            // It may be more practical to use the "index of the
-                            // step ID in the inputs" which would be more stable
-                            // as we change step IDs.
-                            smallvec![ScalarValue::Utf8(Some(format!("step_{}", arg.step_id)))],
-                            smallvec![egg::Id::from(0)],
-                            data_type,
-                        )?;
 
-                        // Then add the actual expression, replacing the `?input` with the fieldref.
-                        let mut subst = egg::Subst::with_capacity(1);
-                        subst.insert(*crate::exprs::INPUT_VAR, egg::Id::from(1));
-                        exprs.add_pattern(&arg.expr, &subst)?;
+                        if let Some(input_step) = arg.step_id {
+                            exprs.add_instruction(
+                                "fieldref",
+                                // Note: that `merge` should produce a record with a
+                                // field for each merged step, identified by the
+                                // step ID being merged. This may change depending
+                                // on (a) optimizations that change steps and
+                                // whether it is difficult to update these and (b)
+                                // how we choose to implement merge.
+                                //
+                                // It may be more practical to use the "index of the
+                                // step ID in the inputs" which would be more stable
+                                // as we change step IDs.
+                                smallvec![ScalarValue::Utf8(Some(format!("step_{}", input_step)))],
+                                smallvec![egg::Id::from(0)],
```

Comment:
Basically, we can choose to use the step ID or the input index. The step ID is "nice" because it is stable across input re-orderings, or if we were to combine merges. On the other hand, the "input index" is nice, because it is stable as we rewrite steps and change step IDs. I suspect we'll get more benefits from using the input index here, but we can adapt that later.

And yeah, `Id::from` is an argument that references the first expression is the input (since we used `ExprPattern::new_input()`).